packages/web3-eth/src/eth_tx.ts

```diff
@@ -0,0 +1,315 @@
+	// TODO - Add default to Web3Context
+	if (populatedTransaction.chain === undefined) populatedTransaction.chain = 'mainnet';
+	// TODO - Add default to Web3Context
+	if (populatedTransaction.hardfork === undefined) populatedTransaction.hardfork = 'london';
+
+	if (populatedTransaction.chainId === undefined) {
+		if (populatedTransaction.common?.customChain.chainId === undefined) {
+			// TODO - web3Eth.getChainId not implemented
+			// populatedTransaction.chainId = await web3Eth.getChainId();
+		}
+	}
+
+	if (populatedTransaction.gas === undefined) {
+		if (populatedTransaction.gasLimit !== undefined)
+			populatedTransaction.gas = populatedTransaction.gasLimit;
+	}
+
+	if (populatedTransaction.gasLimit === undefined) {
+		if (populatedTransaction.gas !== undefined)
+			populatedTransaction.gasLimit = populatedTransaction.gas;
+	}
+
+	populatedTransaction.type = detectTransactionType(populatedTransaction);
+	// TODO - After web3Context.defaultTxType is implemented
+	if (populatedTransaction.type === undefined) populatedTransaction.type = '0x0'; // web3Context.defaultTxType;
+
+	const block = await web3Eth.getBlock();
+	const hexTxType = toHex(populatedTransaction.type);
+
+	if (hexTxType < '0x0' || hexTxType > '0x2')
+		throw new UnsupportedTransactionTypeError(populatedTransaction.type);
+
+	if (hexTxType === '0x0' || hexTxType === '0x1') {
+		// transaction.type not supported before Berlin hardfork
+		// TODO - Maybe add check for populatedTransaction.hardfork >= Berlin before deleting
+		// if (hexTxType === '0x0') populatedTransaction.type = undefined;
+
+		if (populatedTransaction.gasPrice === undefined)
+			populatedTransaction.gasPrice = await web3Eth.getGasPrice();
+	}
+
+	if (hexTxType === '0x1' || hexTxType === '0x2') {
+		if (populatedTransaction.accessList === undefined) populatedTransaction.accessList = [];
+	}
+
+	if (hexTxType === '0x2') {
+		// Unless otherwise specified by web3Context.defaultBlock, this defaults to latest
+		if (block.baseFeePerGas === undefined) throw new Eip1559NotSupportedError();
+
+		if (populatedTransaction.gasPrice !== undefined) {
+			// Logic from 1.x
+			populatedTransaction.maxPriorityFeePerGas = populatedTransaction.gasPrice;
+			populatedTransaction.maxFeePerGas = populatedTransaction.gasPrice;
+			populatedTransaction.gasPrice = undefined;
+		} else {
+			if (populatedTransaction.maxPriorityFeePerGas === undefined)
+				// TODO - Add maxPriorityFeePerGas default to Web3Context
+				populatedTransaction.maxPriorityFeePerGas = toHex(2500000000); // 2.5 Gwei
+			if (populatedTransaction.maxFeePerGas === undefined)
+				populatedTransaction.maxFeePerGas =
+					BigInt(block.baseFeePerGas) * BigInt(2) +
+					BigInt(populatedTransaction.maxPriorityFeePerGas);
+		}
+	}
+
+	// TODO - TSC returns that as only PopulatedUnsignedBaseTransaction
+	return formatTransaction(
+		populatedTransaction,
+		desiredType,
+	) as PopulatedUnsignedTransaction<ReturnType>;
```

Comment:
@nazarhussain Any ideas on what to do about this? The return type should be:

```typescript
export type PopulatedUnsignedTransaction<NumberType = Numbers> =
	| PopulatedUnsignedBaseTransaction<NumberType>
	| PopulatedUnsignedEip2930Transaction<NumberType>
	| PopulatedUnsignedEip1559Transaction<NumberType>;
```

but in my testing the IDE only picks up `PopulatedUnsignedBaseTransaction<NumberType>` and you have to do thing like:

```typescript
const result = await populateTransaction(input, web3Context, ValidTypes.HexString);
			expect((result as PopulatedUnsignedEip1559Transaction).maxPriorityFeePerGas).toBe(
				expectedGasPrice,
			);
```

for TSC/ide not to complain about accessing properties that don't exist on `PopulatedUnsignedBaseTransaction`

Reply:
The return type `PopulatedUnsignedTransaction` is a union. If a function return type is set that, then return value could be any of union types. Users has to use `as` to cast to a concrete type before using it. 

Ideally the function `populateTransaction` should accept a generic which can be inferred from the parameter as well then return one concrete type. 

```ts
onst result = await populateTransaction<PopulatedUnsignedEip1559Transaction>(input, web3Context, ValidTypes.HexString);

// Now use can use it without type cast. 
result.maxPriorityFeePerGas
```

If we can somehow extract this generic type from any of the parameters `input`, `web3Context` then user don't need to pass the generic, but generic still be defined with the function.